x86/defconfig: Add kata containers defconfig

devimc opened the pull request:
kata_containers_common_defconfig is based on Clear Containers config
https://github.com/clearcontainers/packaging/blob/master/kernel/kernel-config-4.9.x
and was adapted to the 4.14.13 kernel by just accepting the defaults configurations

For example, to compile a x86 kernel with KVM enabled:

```
cat arch/x86/configs/kata_containers_common_defconfig arch/x86/configs/kata_containers_kvm_defconfig > arch/x86/configs/kata_containers_defconfig
make ARCH=x86 kata_containers_defconfig
make -j8
```

and to compile a x86 kernel with XEN enabled:

```
cat arch/x86/configs/kata_containers_common_defconfig arch/x86/configs/kata_containers_xen_defconfig > arch/x86/configs/kata_containers_defconfig
make ARCH=x86 kata_containers_defconfig
make -j8
```

fixes #3

Signed-off-by: Julio Montes <julio.montes@intel.com>

grahamwhaley commented:
Hi @devimc - as 2000+ lines of kernel config is quite tricky and takes some time to review ;-), can you give us a little background:
- is this derived from the Clear Containers https://github.com/clearcontainers/linux/commit/9a69950e95b80e07d31e5c7b6227bda4cc87b326 for instance?
- did we adapt to the 4.13.3 kernel by just accepting the defaults for all the new CONFIG entries?

Also, I think you may have submitted this PR against `master` - should it be against/towards `linux-kata-containers-4.13.3` instead?

@gnawux - if you have a kernel config expert, please point /cc them in here - we'd welcome input to share kernel config knowledge (but I also understand it can be quite a task to do a config compare or review, so maybe we will have to schedule that as a task for later).

Generally though, presuming we have tested this kernel in some form, and we get this PR aimed at the correct branch:
lgtm

devimc commented:
Hi @grahamwhaley 
oops! changing to linux-kata-containers-4.13.3

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+CONFIG_CGROUP_PERF=y
+# CONFIG_CGROUP_DEBUG is not set
+CONFIG_SOCK_CGROUP_DATA=y
+CONFIG_CHECKPOINT_RESTORE=y
```
why do we need checkpoint/restore?

grahamwhaley replied:
@devimc, did we need this as part of the investigation of implementing `docker checkpoint` using CRIU?
https://docs.docker.com/engine/reference/commandline/checkpoint/
https://github.com/clearcontainers/runtime/blob/master/docs/limitations.md#checkpointandrestore
https://criu.org/Main_Page

sameo replied:
I actually don't think we need it.

devimc replied:
@grahamwhaley we have to options to support live migration:
1. using libcontainer to migrate only the process
2. using qemu micro checkpoint to migrate all the VM

probably both options won't work because of 9p gaps

devimc replied:
anyway, I'll disable it

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+CONFIG_NET_NS=y
+CONFIG_SCHED_AUTOGROUP=y
+# CONFIG_SYSFS_DEPRECATED is not set
+CONFIG_RELAY=y
```
and relay? Do you want debugfs?

grahamwhaley replied:
I suspect we don't need this - anybody know for sure?

sameo replied:
We don explicitly need it, but it may come as a dependency from something else.

devimc replied:
I'll try to disable it

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+CONFIG_SCHED_AUTOGROUP=y
+# CONFIG_SYSFS_DEPRECATED is not set
+CONFIG_RELAY=y
+# CONFIG_BLK_DEV_INITRD is not set
```
Can you please enable CONFIG_BLK_DEV_INITRD support? We want to use initramfs.

sameo replied:
Yes, @devimc please enable INITRD.

devimc replied:
yes

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+CONFIG_SLABINFO=y
+CONFIG_RT_MUTEXES=y
+CONFIG_BASE_SMALL=0
+# CONFIG_MODULES is not set
```
Can you please enable module support? We'd like to use loadable modules to keep the kernel small.

grahamwhaley replied:
Ah, here we can have an interesting discussion :-)
We have deliberately not enabled modules before for a couple of reasons:
 - size, speed, probing etc. 
 - where will you store the modules (in the container images or on a volume??), and how will you ensure the modules that live on the host match the kernel that you are running inside the VM?

We have debated this a few times. Quite often a new network or filesystem or feature will be needed by a specific client, and having modules available would be one way to fix that.

If we turn on modules, then we should:
- assess any size/speed impact
- I believe turn on kernel API checking so we can try to avoid loading a module/kernel that have a mismatch in API
- discuss signing of modules as well?

sameo replied:
@grahamwhaley All fair points, that should be handled by osbuilder or the kata containers user tools for building its image. Provided that kernel modules support does not impact boot time for kernels that do not load any module at all, I think we should enable it.

bergwolf replied:
Currently we store modules in sandbox rootfs and bind mount it to container rootfs images. Modules are built together with the guest kernel and compressed in the initramfs.

Optional features are built as modules. A sample list includes crypto, nfs, vsock, ipv6, l2tp etc. They are not essential to run a pod and are not loaded if unneeded.

And providing loadable modules opens up the opportunity to provide guest modules if the default kernel config misses something important to a user. In such case, we only need to provide additional kernel modules rather than rebuilding entire kernel.

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+CONFIG_PARAVIRT=y
+# CONFIG_PARAVIRT_DEBUG is not set
+# CONFIG_PARAVIRT_SPINLOCKS is not set
+# CONFIG_XEN is not set
```
We'd like to have XEN PV guest support.

devimc replied:
ok :)

bergwolf commented on arch/x86/configs/kata_containers_defconfig:
```diff
@@ -0,0 +1,2624 @@
+# CONFIG_NET_POLL_CONTROLLER is not set
+CONFIG_TUN=y
+# CONFIG_TUN_VNET_CROSS_LE is not set
+# CONFIG_VETH is not set
```
we need CONFIG_VETH.

sameo replied:
@bergwolf You need `veth` in the guest?

bergwolf replied:
@sameo not for the common case. But we got quite a few `docker in hyperd` and `hyperd in hyperd` requests, which is why we enabled KVM/cgroups support in guest kernel as well.

sameo replied:
@bergwolf Fair enough, thanks.

bergwolf reviewed (commented):
The list is quite long to review. I might have missed something but I already see following differences between cc and hyper kernels:
1. initramfs support
2. loadable module support
3. XEN PV guset
4. veth support

We need to get consensus on them before this can be merged.

laijs commented:
Since we have https://github.com/kata-containers/osbuilder, IMO, it would be better to maintain the kernel config in it.

grahamwhaley commented:
@laijs that's a good question. My only concern would be separating the config away from any patches we are carrying. Having said that, if we end up supporting multiple configs for different performance characteristic/feature sets for instance, then maybe holding those in `osbuilder` makes sense.
/cc @jcvenegas

sameo commented:
@laijs I believe we should keep a reference config file in the kernel and maintain variants in osbuilder, as you suggest.
It should not be mandatory to use osbuilder to build a kata containers kernel, imho.

laijs commented:
@grahamwhaley @sameo Thanks, I got it.

teawater commented:
Hi @devimc ,

We need CONFIG_VMAP_STACK.

Thanks,
Hui

teawater commented:
Hi @devimc ,

I think the kata config should not be included in the kernel repo because I think the kernel repo should just keep the patch that we want to upstream it.

Thanks,
Hui

sameo commented:
@teawater
> We need CONFIG_VMAP_STACK.

Have you benchmarked this? Having stacked vmalloc'ed can potentially introduce TLB misses from stack accesses and could hurt performance.

cc @mcastelino

devimc commented:
after apply the changes suggested by @bergwolf:
```
CONFIG_CHECKPOINT_RESTORE is not set
CONFIG_RELAY is not set
CONFIG_KPROBES is not set
CONFIG_BLK_DEV_INITRD=y
CONFIG_RD_GZIP=y
CONFIG_RD_BZIP2=y
CONFIG_RD_LZMA=y
CONFIG_RD_XZ=y
CONFIG_RD_LZO=y
CONFIG_RD_LZ4=y
CONFIG_STRICT_MODULE_RWX=y
CONFIG_MODULES=y
CONFIG_XEN=y
CONFIG_XEN_PV=y
CONFIG_XEN_PV_SMP=y
CONFIG_XEN_DOM0=y
CONFIG_XEN_PVHVM=y
CONFIG_XEN_PVHVM_SMP=y
CONFIG_XEN_512GB=y
CONFIG_XEN_SAVE_RESTORE=y
CONFIG_VETH=y
CONFIG_HIBERNATE_CALLBACKS=y
CONFIG_PM_SLEEP=y
CONFIG_PM_SLEEP_SMP=y
CONFIG_PM=y
CONFIG_PM_CLK=y
CONFIG_ACPI_TABLE_UPGRADE=y
CONFIG_PCIE_PME=y
CONFIG_XEN_PCIDEV_FRONTEND=y
CONFIG_SYS_HYPERVISOR=y
CONFIG_XEN_BLKDEV_FRONTEND=y
CONFIG_XEN_NETDEV_FRONTEND=y
CONFIG_INPUT_XEN_KBDDEV_FRONTEND=y
CONFIG_HVC_IRQ=y
CONFIG_HVC_XEN=y
CONFIG_HVC_XEN_FRONTEND=y
CONFIG_VT_CONSOLE_SLEEP=y
CONFIG_ZLIB_INFLATE=y
CONFIG_LZO_DECOMPRESS=y
CONFIG_LZ4_DECOMPRESS=y
CONFIG_DECOMPRESS_GZIP=y
CONFIG_DECOMPRESS_BZIP2=y
CONFIG_DECOMPRESS_LZMA=y
CONFIG_DECOMPRESS_XZ=y
CONFIG_DECOMPRESS_LZO=y
CONFIG_DECOMPRESS_LZ4=y
#
# Xen driver support
#
CONFIG_XEN_BALLOON=y
CONFIG_XEN_SCRUB_PAGES=y
CONFIG_XEN_DEV_EVTCHN=y
CONFIG_XEN_BACKEND=y
CONFIG_XENFS=y
CONFIG_XEN_COMPAT_XENFS=y
CONFIG_XEN_SYS_HYPERVISOR=y
CONFIG_XEN_XENBUS_FRONTEND=y
CONFIG_XEN_GNTDEV=m
CONFIG_XEN_GRANT_DEV_ALLOC=m
CONFIG_SWIOTLB_XEN=y
CONFIG_XEN_PCIDEV_BACKEND=m
CONFIG_XEN_PRIVCMD=y
CONFIG_XEN_ACPI_PROCESSOR=m
CONFIG_XEN_HAVE_PVMMU=y
CONFIG_XEN_AUTO_XLATE=y
CONFIG_XEN_ACPI=y
CONFIG_XEN_SYMS=y
CONFIG_XEN_HAVE_VPMU=y
```

teawater commented:
@sameo ,
After open vmap stack, the execution time of test with sysbench memory was decreased about 7% in my part.
And we have long history to use it anti long time execution fragmentation in low memory environment.

Thanks,
Hui

mcastelino commented:
@teawater with vmap stack the performance degrades then. Did you check the kvm stats to see if you are seeing much higher ept violations?

teawater commented:
Sorry to make a mistake in language.  I want to say performance increased but input the execution time increased.

This the clear test result of sysbench memory inside a qemu:
Without vmap stack:
32.6117
31.169
29.1881
30.4291
31.5273

With vmap stack:
29.4660
29.2134
28.0476
28.9640
28.2300

teawater commented:
Following part is what we need:
CONFIG_VMAP_STACK
CONFIG_SERIAL_EARLYCON=y
CONFIG_SERIAL_8250=y
CONFIG_SERIAL_8250_CONSOLE=y
CONFIG_SERIAL_CORE=y
CONFIG_SERIAL_CORE_CONSOLE=y
CONFIG_SCSI=y
CONFIG_SCSI_DMA=y
CONFIG_SG_POOL=y
CONFIG_SCSI_LOWLEVEL=y
CONFIG_SCSI_VIRTIO=y
CONFIG_BLK_DEV_SD=y
CONFIG_CHR_DEV_SG=y
CONFIG_NFS_FS=m
CONFIG_NFS_V2=m
CONFIG_NFS_V3=m
CONFIG_NFS_V4=m

Thanks,
Hui

devimc commented:
@sameo @mcastelino @bergwolf  are you agree with @teawater 's changes ?

devimc commented:
Update
```
CONFIG_NR_CPUS=255
```

sameo commented:
@teawater Why do you need all the serial options? Same question for NFS.

mcastelino commented:
The SCSI changes make sense for sure. We plan to move to SCSI for all block devices. The VMAP change also makes sense as the data from @teawater seems to indicate that this does not degrade performance. It will give us some density benifits. However we still need to run I/O benchmarks.

@grahamwhaley @GabyCT we should run PnP with VMAP enabled for storage and network I/O. We really care about I/O performance more than anything else.

teawater commented:
@sameo We use these kernel options with hyperstart.

Thanks,
Hui

sameo commented:
@teawater I don't think this will be needed with the Kata agent. Could you please double check?

bergwolf commented:
@sameo Yes, we are sure we need them. hyperstart supports NFS volumes. We'd expect the kata agent to support NFS volumes as well. And we need serial ports in case vsock is unavailable, right?

sameo commented:
@bergwolf Yes, I figured you'd need NFS volumes, and that is fine with me, especially since it's going to be a module.
But for serial, the only thing we need is virtio console and tty.

bergwolf commented:
@sameo virtio only works for qemu.

grahamwhaley commented:
Hi @teawater - sorry for the delay here - I was trying to reproduce your VMAP performance changes - can you give me the exact `sysbench` line you are executing when you see that 7% change? Thanks!

mcastelino commented:
> @sameo virtio only works for qemu.

@bergwolf I thought we decided for the first cut we will create a QEMU/KVM centric kernel configuration and more configurations later. 

The question is do we want a single kernel for all hypervisors or have variants for hypervisors. I am ok either way provided we have the data on additional memory consumption (and maybe even boot latency numbers).

teawater commented:
@grahamwhaley 
sysbench  --test=memory run

bergwolf commented:
@mcastelino the config file is called `kata_containers_defconfig` rather than `kata_containers_config_qemu`, and all the Xen related configs are enabled.

While I agree having a per architecture kernel config is good, maintaining a per hypervisor kernel config is too much burden - we end up having to maintain different guest kernels (IOW, rootfs images or initrds) for every architecture and every hypervisor.

teawater commented:
Hi,

I did some code read work about the VMAP_STACK.

About why I got performance increased with VMAP_STACK:
This the comments of Linux kernel:
/*
 * vmalloc() is a bit slow, and calling vfree() enough times will force a TLB
 * flush.  Try to minimize the number of calls by caching stacks.
 */
It makes kernel with VMAP_STACK allocate task stack faster than without VMAP_STACK sometimes.

And I think the more important thing is which type of memory access speed is faster? Get from alloc_pages or get from vmalloc.
I think The access speed of both is very close because they all have EPT support in KVM.
But vmalloc need more TLB entries.

So maybe you can do some tests about task stack allocate speed and task stack access speed inside the Kernel.
Then We can make the appropriate choice based on these test results.

Thanks,
Hui

grahamwhaley commented:
@teawater Ah, great. There are things there we can try. It might be some of the `lmbench` micro-benchmarks (like task creation time etc.) might show us something.
I will try to get to this tomorrow.

sameo commented:
Revamping that one...

@bergwolf There's not much choice for architectures: we must have specific config files for each one of them. However, having a multi hypervisor kernel config is a waste of resources imho, because in several cases we're going to have to enable pieces of code that are exclusive to a given hypervisor. The serial case is a good example. Maybe we could have a common config file with all common options for both Xen and QEMU, and then a config overlay for hypervisor specific pieces. That way we can have hypervisor specific kernels while lowering the maintenance overhead to a minimum.

bergwolf commented:
@sameo I agree with you.

egernst commented:
revamping again.  Should this be rebased (we are not at 4.14, not 4.13).  Also, I believe we agreed that VMAP should be enabled (current patch doesn't enable this).

@bergwolf @sameo  - are we in agreement to create a general defconfig and then apply a hypervisor diffconfig on top of this?

@devimc -- if so, can you work to address (or should we open a new PR)?

grahamwhaley commented:
From my metrics pov, I did try lmbench to see if the microbenchmarks would show me any impacts (good or bad) of the VMAP_STACK. lmbench is a bit noisy, but I could not find any conclusive evidence either way - given that I didn't see major performance drops, I think should go ahead and apply VMAP_STACK to the config.

Commit:
x86/defconfig: Add kata containers defconfigs

kata_containers_common_defconfig is based on Clear Containers config
https://github.com/clearcontainers/packaging/blob/master/kernel/kernel-config-4.9.x
and was adapted to the 4.14.13 kernel by just accepting the defaults configurations

For example, to compile a x86 kernel with KVM enabled:

```
cat arch/x86/configs/kata_containers_common_defconfig arch/x86/configs/kata_containers_kvm_defconfig > arch/x86/configs/kata_containers_defconfig
make ARCH=x86 kata_containers_defconfig
make -j8
```

and to compile a x86 kernel with XEN enabled:

```
cat arch/x86/configs/kata_containers_common_defconfig arch/x86/configs/kata_containers_xen_defconfig > arch/x86/configs/kata_containers_defconfig
make ARCH=x86 kata_containers_defconfig
make -j8
```

fixes #3

Signed-off-by: Julio Montes <julio.montes@intel.com>

devimc commented:
@sameo  @bergwolf  changes applied

sameo commented:
@egernst Yes, that's the idea.

sameo commented:
@devimc Looks good. Maybe we should have a script to safely generate the final Kata config?
@bergwolf @egernst WDYT?

jodh-intel commented:
@sameo - I think that's a **very** good idea. I've raised https://github.com/kata-containers/packaging/issues/8 (and https://github.com/kata-containers/packaging/issues/7 for qemu config).

/cc @grahamwhaley.

devimc commented:
@sameo a script in packaging or here?

jodh-intel commented:
Personally, I think such scripts should live in the packaging repo as that's where they will be used. It also allows repos like this (and qemu) to remain as "vanilla" as possible.

sameo commented:
@devimc @jodh-intel In packaging, yes. Also, see:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/kconfig/merge_config.sh

devimc commented:
@kata-containers/linux  can we merge this PR ?

grahamwhaley commented:
Looks like this is mergeable @devimc , althought the discussion about having a config generating script in the packaging did not seem to conclude.
This PR is also very old - is this config file upto date - I suspect it could do with a re-generation before a merge maybe??

jcvenegas commented:
Any update on this?

jcvenegas commented:
btw I would be good to have the steps described in this PR documented somewhere.

```
cat arch/x86/configs/kata_containers_common_defconfig arch/x86/configs/kata_containers_kvm_defconfig > arch/x86/configs/kata_containers_defconfig
make ARCH=x86 kata_containers_defconfig
make -j8
```

jodh-intel commented:
tbh, I think we could start by simply having the kernel config in the packaging area, as we did for Clear Containers:

- https://github.com/clearcontainers/packaging/blob/master/kernel/kernel-config-4.14.x

WeiZhang555 commented:
I am thinking if it's better to put the kernel config in kernel repo, and qemu config in qemu repo. this can allow us compile our own kernel/qemu without packaging repo.
Seperating kernel repo with kernel config can bring confusion to developers.

jodh-intel commented:
I can understand that viewpoint too. But I think as long as we clearly document the process, either would work. It's worth pointing out that we already have precedent for the hypervisor though:

 -https://github.com/kata-containers/packaging/blob/master/scripts/configure-hypervisor.sh

gnawux commented:
@WeiZhang555 

> I am thinking if it's better to put the kernel config in kernel repo, and qemu config in qemu repo. this can allow us compile our own kernel/qemu without packaging repo.

That's reasonable, however, as linux and qemu are forked from upstream, we'd better keep the stuff not tend to be upstream outside the repo. A packaging script works too as mentioned by @jodh-intel .

WeiZhang555 commented:
@gnawux @jodh-intel That's fine. I believe we can eliminate the confusion by providing some good documentations :smile:

jcvenegas commented:
@grahamwhaley @jodh-intel @gnawux @laijs  what about the proposed PR by @devimc  in https://github.com/kata-containers/packaging/pull/18

jodh-intel commented:
Thanks for the pointer @jcvenegas.

@devimc - do we need to add the `do-not-merge` label to this PR then?
